### PR TITLE
Fix corner case with torch.multinomial

### LINKED ISSUE #4858: CUDA multinomial with replacement can select zero-probability events
I'm running Ubuntu 16.04.3 on an AWS P3.2xlarge, with the NVIDIA libcuda1-384 package, version 384.111-0ubuntu0.16.04.1, and pytorch version 0.3.0.post4 and cuda90, installed via conda.

In some circumstances, the CUDA multinomial sampler with `replacement=True` can sample an event with zero probability.

The script below reproduces this. I haven't tried to reduce the distribution to a minimal case.  The output is below the script.

I've attached [the RNG state](https://github.com/pytorch/pytorch/files/1665704/failed_state.pt.gz) just prior to the failure. Reproduction from this state is demonstrated in the script.

Bests regards,
Alex

```python
import torch as T
import os

freqs = T.cuda.FloatTensor([
    0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
    0.03178183361887932, 0.027680952101945877, 0.033176131546497345,
    0.046052902936935425, 0.07742464542388916, 0.11543981730937958,
    0.14148041605949402, 0.15784293413162231, 0.13180233538150787,
    0.08271478116512299, 0.049702685326337814, 0.027557924389839172,
    0.018125897273421288, 0.011851548217236996, 0.010252203792333603,
    0.007422595750540495, 0.005372154992073774, 0.0045109698548913,
    0.0036087757907807827, 0.0035267581697553396, 0.0018864056328311563,
    0.0024605290964245796, 0.0022964938543736935, 0.0018453967059031129,
    0.0010662291897460818, 0.0009842115687206388, 0.00045109697384759784,
    0.0007791675161570311, 0.00020504408166743815, 0.00020504408166743815,
    0.00020504408166743815, 0.00012302644609007984, 0.0,
    0.00012302644609007984, 4.100881778867915e-05, 0.0, 0.0, 0.0, 0.0,
    0.0, 0.0])

for i in range(1_000_000_000):
    state = T.cuda.get_rng_state()
    sample = T.multinomial(freqs, 1000, True)
    if freqs[sample].min() == 0:
        break

print(72*'-')
print(f'failure after {i} iterations')
sample_idx = (freqs[sample] == 0).nonzero()[0][0]
sampled = sample[sample_idx]
print(f'{sample_idx}th element of last sample was {sampled}, '
      f'which has probability {freqs[sampled]}')

print(72*'-')
SAVE_PATH = 'failed_state.pt'
T.save(state, open(SAVE_PATH, 'wb'))
print(f'\nRNG state saved to `{SAVE_PATH}`')

T.cuda.set_rng_state(T.load(open(SAVE_PATH, 'rb')))
assert freqs[T.multinomial(freqs, 1000, True)].min() == 0

print(72*'-')
print(f'\npytorch version: {T.__version__}\n')
print(72*'-')
print('Operating system:')
os.system('lsb_release -a')
print(72*'-')
print('\nGPU information:')
os.system('nvidia-smi')
print(72*'-')
print('\nCUDA library information:')
os.system('dpkg -l | grep -i cuda')
```

```
ubuntu@ip-172-30-1-60:~/workdir$ python -i fail.py 
------------------------------------------------------------------------
failure after 4309 iterations
631th element of last sample was 0, which has probability 0.0
------------------------------------------------------------------------

RNG state saved to `failed_state.pt`
------------------------------------------------------------------------

pytorch version: 0.3.0.post4

------------------------------------------------------------------------
Operating system:
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.3 LTS
Release:        16.04
Codename:       xenial
------------------------------------------------------------------------

GPU information:
Thu Jan 25 21:17:03 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.111                Driver Version: 384.111                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   38C    P0    46W / 300W |   4558MiB / 16152MiB |     31%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0       988      C   python                                       738MiB |
|    0     15193      C   python                                      3810MiB |
+-----------------------------------------------------------------------------+
------------------------------------------------------------------------

CUDA library information:
ii  libcuda1-384                                    384.111-0ubuntu0.16.04.1     amd64                        NVIDIA CUDA runtime library
```

### PR DESCRIPTION
**Pull request by t-vi**
In the shortcut for n_sample=1, when category 0 has 0 weight,
we should not map the (uniform) sample 0 to category 0.
The conversion uniform->multinomial was apparently written to work on
a (0,1] range (like curand uses), but PyTorch uses a [0,1) range.

Fixes: #4858. Thank you, Roy Fejgin for reporting.

### CODE REVIEW

**Comment by t-vi** [comment]
Fair enough, I've changed the fix to substitute 0 for 1 (mirroring the move from curand -> PyTorch uniform) and added a comment. Quite likely, one might rewrite the function a bit to work with [0,1) when it is migrated to ATen native or so.

**Comment by rfejgin** [comment]
Thanks @t-vi . Do you think the second test in this line still makes sense after this fix?
https://github.com/t-vi/pytorch/blob/f6eac1b3cfffba97ce5d3ce2d429fbf22a678151/aten/src/THC/THCTensorRandom.cuh#L215

It seems like that second condition can now never be true. Also, if the code was originally written to work with samples in the interval (0,1] I wonder why it's testing for sample==0?

**Comment by t-vi** [comment]
Ha. That was exactly the test I had missed. Thanks Roy!
So before I had moved to fix it the way I did now (for which I checked the history of the code), I tried to adjust the code below to handle 0 correctly. Now, I had missed this part and the continue here causes the category to be 0, and my tested adjustment to the binning intervals didn't work out.
I'll revisit that plan for a clean fix. Thanks for pointing it out!

**Comment by t-vi** [comment]
Thanks! You could just capture and save the RNG state to a temporary variable (torch.cuda.get_rng_state) and output it after detecting failure.

**Comment by rfejgin** [comment]
Here are the probabilities. Will post a script shortly.

[weights.zip](https://github.com/pytorch/pytorch/files/2247080/weights.zip)

**Comment by rfejgin** [comment]
Attaching RNG state
[rng_state.zip](https://github.com/pytorch/pytorch/files/2247082/rng_state.zip)

**Comment by rfejgin** [comment]
Here is my script. I have not had a chance to create an absolute minimal test case, but if you run this you should see the problem when it gets iteration n 61678, which takes about 52 seconds on my GPU (1080ti).
[debug_zero_prob.zip](https://github.com/pytorch/pytorch/files/2247108/debug_zero_prob.zip)

**Comment by rfejgin** [comment]
Attaching a new version of the test case and output. This version has slightly clearer printouts.

[debug_zero_prob.zip](https://github.com/pytorch/pytorch/files/2247444/debug_zero_prob.zip)

**Comment by rfejgin** [comment]
The strange thing is that now, these zero probability bins getting selected are not bin zero, nor bins that are near any bin that has non-zero probability. For example the PDF can look like this:
[0.0, 0.0, 0.0, ... 0.0, 0.3, 0.4, 0.1, 0.0, 0.0, 0.0, ...  0.0, 0.0, **0.0**, 0.0, 0.0, 0.0]
.. and the bold bin gets selected. How is that bin different from any of its neighboring bins? The CDF should have the same value for this bin and its neighbors, so seems to me that it can't be a subtlety in the bucket boundary checks.
I'm starting to suspect something more fundamental is wrong with the CUDA implementation... but just speculating.

**Comment by rfejgin** [comment]
I added some tracing in the CUDA code. In the case I caught, the uniform sample equals 0.99999994. Then many buckets (to the right of the last non-zero-probability bin) ended up meeting the condition

 ` (cat < categories) &&
        (!THCNumerics<T>::ge(sample, curBucket)) &&
        (THCNumerics<T>::ge(sample, prevBucket));`

All of them had curBucket == 1.0 and prevBucket == 0.99999994

Then many threads end up writing out the result in this line
`dest[curDist] = cat + TH_INDEX_BASE;`

and I think that maybe the last of the threads to make that update wins.

 
Now I guess the question is why curBucket does not equal prevBucket even though all these bins have zero probability.

**Comment by rfejgin** [comment]
Oops, I just realize you must have wanted the RNG state *before* the bug occurs so that you could regenerate the problematic uniform samples.What I posted was the RNG state after the bug occurred -sorry, I misunderstood. I'm not at my desk now to re-capture that data, but hopefully the test case I posted and the uniform sample value is enough to analyze / reproduce this.

**Comment by t-vi** [comment]
I hope to look into it later today.

**Comment by t-vi** [comment]
So I took the liberty to condense the test case to
```
batch_size = 1024
dist_size = 2048
k = 16095
torch.cuda.manual_seed(k)
torch.manual_seed(k)
weights = torch.empty(batch_size, dist_size, device='cuda')
weights.uniform_(0.4, 0.6)
weights[:, :int(dist_size/2)] = 0
weights[:, int(dist_size/2)+4:] = 0
s = weights.multinomial(1).squeeze(1)
selected_probs = weights[torch.arange(batch_size), s]
assert not (selected_probs==0).any()
```

**Comment by t-vi** [comment]
So I took the liberty of condensing your example to the following:
```
import torch
print (torch.__version__)
batch_size = 1024
dist_size = 2048
k = 16095
torch.cuda.manual_seed(k)
torch.manual_seed(k)
weights = torch.empty(batch_size, dist_size, device='cuda')
weights.uniform_(0.4, 0.6)
weights[:, :int(dist_size/2)] = 0
weights[:, int(dist_size/2)+4:] = 0
s = weights.multinomial(1).squeeze(1)
selected_probs = weights[torch.arange(batch_size), s]
assert not (selected_probs==0).any()
```
(and it really fails with e5c5ae302b60339ec50438885060b96479c32df8 , I had just realized I had another branch when I first posted this).

**Comment by t-vi** [comment]
So interestingly, when I use the identical kernel with a caller that calls torch.rand and then the kernel from a PyTorch extension, that seems to not have the error. I'll see to extract the random part from the actual call.
A random thing I've noticed and been wondering is whether there is a typo regarding the shared memory requirements: There is the product [`sizeof(real)*sizeof(accreal)`](https://github.com/pytorch/pytorch/blob/e8f27311aa22dc565fa310d6c820783d9c273cd1/aten/src/THC/generic/THCTensorRandom.cu#L184) looks pretty suspicious and probably should be a sum (possibly with coeffs). But that likely is unrelated to our present bug.

**Comment by rfejgin** [comment]
It's starting to look like a floating point numeric accuracy thing, that the code isn't resilient to. I'm seeing cases where curBucket should be equal to prevBucket (because the bucket itself has zero probabilitly), but they are slightly different, with curBucket being 1.0 and prevBucket being a little less than 1.0. If the sample falls between those values, the bucket gets selected.
I think curBucket prevBucket *should* be equal mathematically but are diverging because of floating point arithmetic -- they are the result of different sums calculated by different threads.

**Comment by rfejgin** [comment]
@t-vi  interesting finding about the memory allocation, btw. Yeah, that product does look strange. Though probably allocating too *much* memory which is better than the opposite... :)

**Comment by rfejgin** [comment]
Here's a log showing an example where bin 1054 was selected. All bins from 1028 onwards had zero probability. Hmm, at least with this example I think the problem would not have happened with if the condition was 'gt' as it originally was. Let's think about whether that check fundamentally makes more sense or just happens to work better in this case.
`curDist=329,thread=1023,cat=1023,prevBucket=0,curBucket=0,sample=0.999999940395355
curDist=329,thread=0,cat=1024,prevBucket=0,curBucket=0.227620840072632,sample=0.999999940395355
curDist=329,thread=1,cat=1025,prevBucket=0.227620840072632,curBucket=0.502622127532959,sample=0.999999940395355
curDist=329,thread=2,cat=1026,prevBucket=0.502622127532959,curBucket=0.769631028175354,sample=0.999999940395355
curDist=329,thread=3,cat=1027,prevBucket=0.769631028175354,curBucket=0.999999940395355,sample=0.999999940395355
curDist=329,thread=4,cat=1028,prevBucket=0.999999940395355,curBucket=1,sample=0.999999940395355
curDist=329,thread=5,cat=1029,prevBucket=1,curBucket=0.999999940395355,sample=0.999999940395355
curDist=329,thread=6,cat=1030,prevBucket=0.999999940395355,curBucket=1,sample=0.999999940395355
`

**Comment by rfejgin** [comment]
Maybe if we want ensure zero-probability events are not selected, in a way that is robust to floating point numeric problems, the `inBucket ` check could be forced to false if the difference between prevBucket and curBucket is less than an epsilon. That epsilon might need to depend on the underlying data type (T) being used.

**Comment by t-vi** [comment]
So I think the root cause is the [inclusive prefix sum (in the comment)](https://github.com/pytorch/pytorch/blob/4a5cd4f6ab2cc9d48663189ec43e8ac20a4bb252/aten/src/THC/THCTensorRandom.cuh#L237). Here, the summation order is totally different for adjacent elements, that might lead to the observed non-monotonicity and other effects.
Now how to fix this? I see the following options:
- There is the minimal size @rfejgin suggests at #4858.
- One elaborate option (which I'm not 100% certain that it will suffice) would be to improve the summing - I tested using the powers of two in the bit pattern as summation units and it looks like it might help.
- Rounding the inputs would also likely work.
- My favourite solution would be to just test dist > 0 in the same place as Roy would check the distances. I implemented that.
A final observation: the prefix sum is done using T instead of AccT. I'm not sure why given that shared memory of AccT type is available but I didn't change it for now.

(I accidentally pushed this before building, sorry.)

**Comment by t-vi** [comment]
So to me it looks like this test case captures the essence of @rfejgin 's latest example and works as expected with the latest commit.
```
import torch
dist_size = 2048
batch_size = 1024
torch.cuda.manual_seed(41134)
weights = torch.zeros(dist_size, device='cuda')
weights[[1024,1025,1026,1027]] = torch.tensor([0.47056111693382263, 0.568510890007019, 0.5519880056381226, 0.476242333650589], device='cuda')
weights = weights[None].expand(batch_size, dist_size)
s = weights.multinomial(1).squeeze(1)
selected_probs = weights[torch.arange(weights.shape[0]), s]
assert not (selected_probs==0).any()
```

If @rfejgin is reasonably happy with this, I'll add the above as a test case and remove the not ready yet label.

**Comment by rfejgin** [comment]
Thanks Thomas. I agree with you on the root cause. I also like the idea of explicitly checking if the distribution value equals zero, and was having similar thoughts after I posted. The explicit check avoids the need to tune a threshold (or one per type of T), which I've actually tried and found fragile. So your fix looks good to me!

One minor comment is that I would consider putting the memory allocation fix in a different commit. In the unlikely case that that change causes instability it would be good to be able to separate the two changes.

**Comment by t-vi** [comment]
OK. My impression is that PRs are heavyhanded enough to tack the shared memory size on this, but I'll gladly remove it if the reviewer prefers.

**Comment by rfejgin** [comment]
@t-vi: it's really up to you - this is my first pytorch PR review, so I'm not super familiar with the conventions - just a suggestion.

On another topic: Now that we're explicitly handling the zero-probability case, do you think that the change in the condition from gt to ge is still necessary? https://github.com/pytorch/pytorch/pull/9960/files#diff-a5d44fe2c50c7cc4616e927154fb7e30R259

**Comment by t-vi** [comment]
You could argue that 0 is exceedingly rare, but I think it is good practice to cover the half-open interval [0,1) with equally half-open bins [a,b), which is what the ge achieves (and the gt seems to stem from the time when the uniform random number was in fact in (0,1]).

**Comment by rfejgin** [comment]
Ah, I see the logic now. Makes sense.

**Comment by rfejgin** [comment]
I just wanted to mention that my availability will be very limited in the next two weeks. But it looks like we're pretty close to resolving this -- nice :).

**Comment by yf225** [comment]
@rfejgin @t-vi What should be the next step for this PR?

**Comment by t-vi** [comment]
I think it's ready for review and possibly merging.

**Review by soumith — APPROVED** [review]
